Sources/SwiftSyntaxMacros/SwiftSyntaxMacros.docc/SwiftSyntaxMacros.md

```diff
@@ -1,37 +1,308 @@
+
+Finally, every syntax type has a memberwise initializer with which a syntax node can be created by specifying all its children. This is the most low-level way of creating syntax nodes and has the most flexibility. It is also a good option for nodes that are not representable by string literals and don’t have result builder initializers.
+
+## Emitting Errors
+
+Macros are not always applicable. For example, it doesn’t make sense to apply the `EnumSubset` macro on a type that is not an enum. If a macro is used in ways that it doesn’t support, it is always better to emit custom error messages that tells the adopter about what’s going wrong, instead of having them read the generated code to debug the macro.
+
+Macro errors can be represented by any type that conforms to the Swift `Error` protocol. For example, an enum with a single case can be used to describe the error message if `EnumSubset` is applied to a type that's not an enum.
+
+```swift
+enum EnumSubsetError: CustomStringConvertible, Error {
+  case onlyApplicableToEnum
+  
+  var description: String {
+    switch self {
+    case .onlyApplicableToEnum: return "@EnumSubset can only be applied to an enum"
+    }
+  }
+}
+```
 
-## What's the future for this library?
+An error can be emitted by throwing it from the expansion function. Its message will be shown inline with any other compilation errors.
 
-Well, that depends. The `SwiftSyntaxMacros` module is part of a potential vision for a macros feature in Swift, where the effect of macro evaluation is as it works here---executable Swift code that transforms a given syntax tree into another syntax tree. That design also integrates macros into the type checker, so that macros are only evaluated on well-typed inputs and are expected to either produce well-typed syntax trees as output or emit diagnostics along the way. If that design is accepted as the design for Swift, much of this module will remain as it is---but the examples themselves will be divided into a builtin set of macros that are always available to the compiler (e.g., to provide the implementation for things like `#line`) or moved into a separate examples module to provide documentation on how to use the feature.
+The ``MacroExpansionContext/addDiagnostics(from:node:)`` method on the ``MacroExpansionContext`` allows further customization of the error message, like showing the error message at a different location than the attribute, generating warnings, or even providing Fix-Its in Xcode.
```

Comment:
> There's two ways to emit this error:
>   1. Throw the error from the expansion function. This will generate a compilation error on the line of the attribute.
>   2. Add it as a diagnostic with  ``MacroExpansionContext/addDiagnostics(from:node:)``. This allows further customization, eg. changing the location, emitting a warning instead of an error, or even providing Fix-Its.

Reply:
Changed to

> There are two ways to emit this error:
>
> 1. Throw an error from the `expansion` function. This generates a compilation error on the line of the attribute. This is the technique that the above example uses. 
> 2. Add it as a diagnostic with  ``MacroExpansionContext/diagnose(_:)``. This allows further customization, like changing the location, emitting a warning instead of an error, or even providing Fix-Its.